src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -220,19 +185,94 @@ public Connection getConnection() {
         return this.subscription.connection;
     }
 
+    @Override
     public String getSubject() {
         return this.subject;
     }
 
+    @Override
     public String getReplyTo() {
         return this.replyTo;
     }
 
+    @Override
+    public byte[] getSerializedHeader() {
+        return headers == null || headers.size() == 0 ? null : headers.getSerialized();
+    }
+
+    @Override
+    public Headers getHeaders() {
+        return headers;
+    }
+
+    @Override
     public byte[] getData() {
-        return this.data.array();
+        return this.data;
     }
 
+    @Override
     public Subscription getSubscription() {
         return this.subscription;
     }
+
+    @Override
+    public String toString() {
+        String hdrString = headers == null ? "" : new String(headers.getSerialized(), US_ASCII).replace("\r", "+").replace("\n", "+");
+        return "NatsMessage:" +
+                "\n  subject='" + subject + '\'' +
+                "\n  replyTo='" + replyTo + '\'' +
+                "\n  data=" + (data == null ? null : new String(data, UTF_8)) +
+                "\n  utf8mode=" + utf8mode +
+                "\n  headers=" + hdrString +
+                "\n  sid='" + sid + '\'' +
+                "\n  protocolLineLength=" + protocolLineLength +
+                "\n  protocolBytes=" + (protocolBytes == null ? null : new String(protocolBytes, UTF_8)) +
+                "\n  kind=" + kind +
+                "\n  hpub=" + hpub +
+                "\n  sizeInBytes=" + sizeInBytes +
+                "\n  hdrLen=" + hdrLen +
+                "\n  dataLen=" + dataLen +
+                "\n  totLen=" + totLen +
+                "\n  subscription=" + subscription +
+                "\n  next=" + next;
+    }
+
+    // The builder is only supports building normal publish/request messages,
+    // as an option for client use developers instead of the normal constructor
+    public static class Builder {
+        String subject;
+        String replyTo;
+        Headers headers;
+        byte[] data;
+        boolean utf8mode;
+
+        public Builder subject(final String subject) {

```

Comment:
We'll want comments on these - I think the gradle build will fail on the javadoc task w/o them.

Reply:
yes will add. Not tonight though. Is there a gradle command I should be running before commit?

Reply:
I do `./gradlew build`

Reply:
the javadoc excludes `excludes = ['io/nats/client/impl', 'io/nats/examples']` but I'll still add comments